### ISSUE TITLE
ConditionEnvironment= not affected by "systemctl set-environment"

### ISSUE THREAD

**Issue by WhyNotHugo** [systemd/systemd#20366]
**systemd version the issue has been seen with**

```
systemd 249 (249.rc1.r113.g409ca5fff1-1-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

**Used distribution**
ArchLinux

**Linux kernel version used** (`uname -a`)

    Linux victory 5.12.14-arch1-1 #1 SMP PREEMPT Thu, 01 Jul 2021 07:26:06 +0000 x86_64 GNU/Linux

**CPU architecture issue was seen on**

amd64

**Steps to reproduce**

I'm using `mako`,  a notification daemon for wayland. It includes a systemd.service(5) file which runs conditionally only if `WAYLAND_DISPLAY` is set:

https://github.com/emersion/mako/blob/master/contrib/systemd/mako.service#L7

If _any_ application tries to send a notification before wayland starts, systemd evaluates the condition and refuses to start mako. This makes perfect sense.

**Unexpected behaviour you saw**

If I later start `sway` (a wayland compositor) and `systemctl --user import-environment WAYLAND_DISPLAY`, systemd will continue to refuse to start `mako`.

**Expected behaviour you didn't see**

If I later start `sway` (a wayland compositor) and `systemctl --user import-environment WAYLAND_DISPLAY`, systemd should start `mako`.

**Additional info**

Just in case:

```console
$ systemctl --user show-environment | grep WAYLAND_DISPLAY
WAYLAND_DISPLAY=wayland-1
```

Apparently the issue is that `systemd` will only evaluate `ConditionEnvironment=` **once**. I'm not sure if this is be design. The docs don't mention that this is the case, so it doesn't sound like it should.

Using `ExecCondition=/bin/sh -c '[ -n "$WAYLAND_DISPLAY" ]'` as a hack does work.

**Comment by poettering** [comment]
ExecCondition= checks the env block of the manager process, i.e. the one the manager process (i.e. "systemd") itself received. This is different from the env block that forked off processes will receive, which is what "import-environment"/"show-environment" operates on.

i.e there's a distinction between what systemd itself operates with and what it passes down to others. 

I am not convinced we should really change this.

**Comment by WhyNotHugo** [comment]
I've no strong opinion either way, but the behaviour is not changed, the docs should be a bit clearer, since both refer to the `manager environment block` ("service manager" here refers to `systemd`, right?):

```
ConditionEnvironment=
   ConditionEnvironment= may be used to check whether a specific environment
   variable is set (or if prefixed with the exclamation mark — unset) in the
   service manager's environment block. [...]


import-environment VARIABLE...
   Import all, one or more environment variables set on the client into the
   systemd manager environment block. [...]
```

That aside, there's a bit of inconsistency in behaviour...

**Scenario 1**

- Run `import-environment`.
- Run the service: Service starts

**Scenario 2**

- Run the service (it doesn't run)
- Run `import-environment`.
- Run the service: Service does not start.

It seem so `ConditionEnvironment=` is evaluated the first time there's an attempt to start the service, but _does_ take into consideration values that were imported.

**Comment by WhyNotHugo** [comment]
Already fixed in #388

**Comment by emersion** [comment]
This PR seems to be unrelated? Is this a typo in the PR number?

**Comment by rpigott** [comment]
> ExecCondition= checks the env block of the manager process

@poettering I don't think that is correct? It's actually pretty well documented in the code otherwise. The simplified call chain is something like:

```
unit_start(unit, ...)
unit_test_condition(unit)
  manager_get_effective_environment(unit->manager, &env)
    *env = strv_env_merge(manager->transient_environment, manager->client_environment);
condition_test_list(..., env, ...)
condition_test(condition, env)
condition_test_environment(condition, env)
  STRV_FOREACH(i, env) { ... }
```

where those environments are documented as:

https://github.com/systemd/systemd/blob/72e7bfe02d7814fff15602726c7218b389324159/src/core/manager.h#L250-L251

And `client_environment` is indeed the one modified by the SetEnvironment method.

In practice I observe behavior in-line with my interpretation above — units with ConditionEnvironment= are passed even when the variable is only present in the client block following modification by `systemctl import-environment`. It can be tested with a basic service like:

```
[Unit]
Description=A nonsense service for testing
ConditionEnvironment=NONSENSE

[Service]
Type=simple
ExecStart=/usr/bin/sleep 1
```

```
systemctl --user start nonsense.service # fails condition check
systemctl --user set-environment NONSENSE=foobar
systemctl --user start nonsense.service # passes condition check
```

I cannot reproduce the behavior in the issue description. Based on a cursory look, it has worked this way since at least 2018, probably before.
